db/db_impl/db_impl.cc

```diff
@@ -1061,6 +1061,204 @@ FSDirectory* DBImpl::GetDataDir(ColumnFamilyData* cfd, size_t path_id) const {
+        uint64_t latest_applied_update_sequence = 0;
+        for (auto& e : edits) {
+          if (!e.HasManifestUpdateSequence()) {
+            return Status::InvalidArgument(
+                "Manifest write doesn't have a ManifestUpdateSequence");
+          }
+          latest_applied_update_sequence = e.GetManifestUpdateSequence();
+          if (e.GetManifestUpdateSequence() <= current_update_sequence) {
+            // Ignore the update, we already have it
+            continue;
+          }
+          ++current_update_sequence;
+          if (e.GetManifestUpdateSequence() != current_update_sequence) {
+            std::ostringstream oss;
+            oss << "Gap in ManifestUpdateSequence, expected="
+                << current_update_sequence
+                << " got=" << e.GetManifestUpdateSequence();
+            return Status::Corruption(oss.str());
+          }
+
+          if (e.HasLogNumber()) {
+            // Physical replication does not support WAL, but RocksDB expects
+            // some invariants of alive_log_files_ and logs_ and who are we to
+            // argue: at least one alive_log_files_ and logs_ always needs to be
+            // alive and alive means that its log number is greater or equal to
+            // the latest log number.
+            versions_->MarkFileNumberUsed(e.GetLogNumber());
+            alive_log_files_.begin()->number = e.GetLogNumber();
+            logs_.front().number = e.GetLogNumber();
+          }
+
+          ColumnFamilyData* cfd{nullptr};
+          if (!e.IsColumnFamilyAdd()) {
+            cfd = versions_->GetColumnFamilySet()->GetColumnFamily(
+                e.GetColumnFamily());
+          }
+          if (e.IsColumnFamilyAdd()) {
+            single_column_family_mode_ = false;
+            added_column_families.push_back(e.GetColumnFamily());
+            cf_options = std::make_unique<ColumnFamilyOptions>(
+                default_cf_handle_->cfd()->GetLatestCFOptions());
+          } else if (e.IsColumnFamilyDrop()) {
+            info->deleted_column_families.push_back(e.GetColumnFamily());
+          }
+          cfds.push_back(cfd);
+          mutable_cf_options_list.push_back(mutable_options);
+          autovector<VersionEdit*> el;
+          el.push_back(&e);
+          edit_lists.push_back(std::move(el));
+        }
+        s = versions_->LogAndApply(cfds, mutable_cf_options_list, edit_lists,
+                                   &mutex_, directories_.GetDbDir(),
+                                   false /* new_descriptor_log */,
+                                   cf_options.get());
+        if (!s.ok()) {
+          return s;
+        }
+        for (auto cfd : cfds) {
+          if (!cfd) {
+            continue;
+          }
+          auto& sv_context = job_context.superversion_contexts.back();
+          cfd->InstallSuperVersion(&sv_context, &mutex_);
+          sv_context.NewSuperVersion();
+          cfd->imm()->RemoveOldMemTables(cfd->GetLogNumber(),
```

Comment:
Uh to be honest I'm not 100% sure why we need new superversion here, I took this from the other usage of `RemoveOldMemtables`: https://github.com/rockset/rocksdb-cloud/blob/11877b32890419c92731fde5aa9213a47689782c/db/db_impl/db_impl_secondary.cc#L598

> Why don't we do the same for mem table switch?

Memtable switch creates its own superversion:

https://github.com/rockset/rocksdb-cloud/blob/11877b32890419c92731fde5aa9213a47689782c/db/db_impl/db_impl_write.cc#L2021

> Also, why do we remove the old mem tables here?

We need to remove old memtables in case the write to manifest was a flush. Flush needs to atomically add an SST file and remove the memtable corresponding to that file. The first part happens in `LogAndApply()` and the second part happens here.